---
title: 'STATS 202 Homework 3'
---

In [143]:
from ISLP import load_data
from sklearn import model_selection
from sklearn.linear_model import LinearRegression, LassoCV, RidgeCV
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.cross_decomposition import PLSRegression
from sklearn.pipeline import make_pipeline
from sklearn import decomposition
from sklearn.preprocessing import StandardScaler
from IPython.display import Markdown
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

sns.set_theme()

# Problem 1
## Problem 1.a
**iv.**: Train RSS will steadly decrease, as the model becomes more flexable and will be able to fit the data better.
## Problem 1.b
**ii.**: Test RSS will decrease as the flexability lets it get closer to the real function $Y_0$, then eventually start overfitting on the training data and therfore having an increased test RSS.
## Problem 1.c
**iii.**: Variance will steadly increase as the model becomes more flexable and varies to the data better
## Problem 1.d
**iv.**: squared bias will steadly decrease as the model fits the data better
## Problem 1.e
**v.**: Irreducible error will remain constant as it is independent of the model.
# Problem 2
## Problem 2.a
**iii.**: Train RSS will steadly increase as the model gets more punished for flexability.
## Problem 2.b
**ii.**: Test RSS decrease as the increased penalty term prevents overfitting, while it eventually restricts the model from representing $Y_0$, therfore increasing test RSS.
## Problem 2.c
**iv.**: Variance will steadly decrease as the model is penalized for varying to fit the data
## Problem 2.d
**iii.**: Squared bias will steadly increase as the penalty term causes the model to be less specific and therfore more biased.s
## Problem 2.e
**v.**: Irreducable error is independent of the model so will be constant.
# Problem 3
## Problem 3.a

In [128]:
College = load_data("College")
College["Private"] = College["Private"] == "Yes"
X_train, X_test, y_train, y_test = model_selection.train_test_split(
    College.drop("Apps", axis=1), College["Apps"], random_state=42
)
scaler = StandardScaler().fit(X_train)
scaled_X_train = scaler.transform(X_train)
scaled_X_test = scaler.transform(X_test)

## Problem 3.b

In [129]:
Markdown(
    "The score is: "
    + str(LinearRegression().fit(X_train, y_train).score(X_test, y_test))
)

The score is: 0.8606501781209516

## Problem 3.c

In [130]:
Markdown(
    "The score is: "
    + str(
        RidgeCV(alphas=np.arange(1, 100))
        .fit(scaled_X_train, y_train)
        .score(scaled_X_test, y_test)
    )
)

The score is: 0.8626496724638395

## Problem 3.d

In [131]:
Markdown(
    "The score is: "
    + str(
        LassoCV(alphas=np.arange(1, 100))
        .fit(scaled_X_train, y_train)
        .score(scaled_X_test, y_test)
    )
)

The score is: 0.8610062021214084

In [132]:
Markdown(
    "the number of nonzero coefficents is: "
    + str(
        np.sum(
            LassoCV(alphas=np.arange(1, 100)).fit(scaled_X_train, y_train).coef_ != 0
        )
    )
)

the number of nonzero coefficents is: 17

## Problem 3.e

In [139]:
grid = GridSearchCV(
    make_pipeline(PCA(), LinearRegression()), {"pca__n_components": np.arange(1, 17)}
).fit(scaled_X_train, y_train)
Markdown("The best M is: " + str(grid.best_params_["pca__n_components"]))

The best M is: 16

In [140]:
Markdown("The score is: " + str(grid.score(scaled_X_test, y_test)))

The score is: 0.8664792682433866

## Problem 3.f

In [145]:
grid = GridSearchCV(PLSRegression(), {"n_components": np.arange(1, 17)}).fit(
    scaled_X_train, y_train
)
Markdown("The best M is: " + str(grid.best_params_["n_components"]))

The best M is: 16

In [146]:
Markdown("The score is: " + str(grid.score(scaled_X_test, y_test)))

The score is: 0.8606488441076391

## Problem 3.g

There is not a significant differnce between the results for these methods; without fruther investigation and much deeper analisys, I don't even think we can conclude that these errors have a statistically significant difference. We see that the most accurate model is PCR ( with a $\approx 87\%$ accuracy ), while the least accurate is the simple linear regression ( with a $\approx 86\%$ accuracy ).

# Problem 4
## Problem 4.a
$$
\begin{split}
f_1(x)&=\beta_0+\beta_1 x+\beta_2 x^2+\beta_3 x^3+\beta_4(x-\xi)^3_+ \\
&=\beta_0+\beta_1 x+\beta_2 x^2+\beta_3 x^3 \\
a_1&=\beta_0 \\
b_1&=\beta_1 \\
c_1&=\beta_2 \\
d_1&=\beta_3+\beta_4
\end{split}
$$
## Problem 4.b
$$
\begin{split}
f_2(x)&=\beta_0+\beta_1 x+\beta_2 x^2+\beta_3 x^3+\beta_4(x-\xi)^3_+ \\
&=\beta_0-\beta_4 \xi^3+x(\beta_1+3\beta_4 \xi^2)+x^2(\beta_2-3\beta_4 \xi)+x^3(\beta_3+\beta_4) \\
a_2&=\beta_0-\beta_4 \xi^3 \\
b_2&=\beta_1+3\beta_4 \xi^2 \\
c_2&=\beta_2-3\beta_4 \xi \\
d_2&=\beta_3+\beta_4
\end{split}
$$
## Problem 4.c
$$
\begin{split}
f_1(\xi)&=\beta_0+\beta_1 \xi+\beta_2 \xi^2+\beta_3 \xi^3 \\
f_2(\xi)&=\beta_0-\beta_4 \xi^3+\beta_1 \xi+3\beta_4 \xi^3+\beta_2 \xi^2-3\beta_4 \xi^3+\beta_3 \xi^3+\beta_4 \xi^3 \\
&=\beta_0+\beta_1 \xi+\beta_2 \xi^2+\beta_3 \xi^3
\end{split}
$$
## Problem 4.d
$$
\begin{split}
f_1'(\xi)&=\beta_1+2\beta_2 \xi+3\beta_3 \xi^2 \\
f_2'(\xi)&=\beta_1+3\beta_4 \xi^2+2\xi(\beta_2-3\beta_4 \xi)+3\xi^2(\beta_3+\beta_4) \\
&=\beta_1+2\beta_2 \xi+3\beta_3 \xi^2
\end{split}
$$
## Problem 4.e
$$
\begin{split}
f_1''(\xi)&=2\beta_2+6\beta_3 \xi \\
f_2''(\xi)&=2(\beta_2-3\beta_4 \xi)+6\xi(\beta_3+\beta_4) \\
&=2\beta_2+6\beta_3 \xi
\end{split}
$$


# Problem 5
Use ANOVA